ENH: fix 0d array printing using `str` or `formatter`.

ahaldane opened the pull request:
This is split off from #9139, plus has some extra editing of the release notes.

ahaldane committed:
ENH: fix 0d array printing using `str` or `formatter`.

The str of 0d arrays now returns `str(a[()])` like scalars, and the
repr returns `'array(' + formatter(a[()]) + ')'` like ndarrays.
The default implementation of str and repr for user-defined types is
removed. Fixes #1415

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -521,7 +515,9 @@ def array2string(a, max_line_width=None, precision=None,
 
-    if a.size == 0:
+    if style is not None and a.shape == () and a.dtype.names is None:
+        return style(a[()])
```
Worth adding a comment here that the `names is None` check is to avoid co-recursion with `voidtype_str`, which upcasts scalars to arrays again

ahaldane commented:
Added comment, also reorganized voidtype_str.

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -621,6 +594,33 @@ static PyObject *
+        return ret;
+    }
+    else {
+        PyObject  *item, *item_str;
```
Too many spaces here

eric-wieser commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -237,9 +237,22 @@ def test_formatter_reset(self):
+        # repr of 0d arrays is affected by printoptions
         x = np.array(1)
         np.set_printoptions(formatter={'all':lambda x: "test"})
         assert_equal(repr(x), "array(test)")
```
Can we add a test here that `str(x)` is not affected? Also, doesn't this pollute all future tests unless we remove the new formatter?

ahaldane replied:
Your comment below got lost on github, but I see it in the email.

This test class implements `tearDown` which resets the printoptions, so we don't need to worry about pollution. Will add the str test.

eric-wieser reviewed (approved):
LGTM. There's some related stuff in #9201, but we can deal with that later.

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -621,6 +594,33 @@ static PyObject *
+            ret = PyObject_Str((PyObject *)arr);
+            Py_DECREF(arr);
+        }
+        return ret;
```
Can we get rid of this, and make a call to something in `core.multiarray._internal`? This is the recursive point I'd rather not have.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -521,7 +515,11 @@ def array2string(a, max_line_width=None, precision=None,
-    if a.size == 0:
+    # the ``dtype.names is None`` check is to avoid co-recursion with
+    # voidtype_str, which upcasts structured scalars to 0d arrays
+    if style is not None and a.shape == () and a.dtype.names is None:
```
Then we can ditch this check too

ahaldane commented:
Yeah, eliminating the recursion that way is a good idea. Done in the new commit.

There was a small catch, though, that we have to treat subarray-fields carefully. Previously they were specially printed using the `SubarrayFormatter` which leaves out newlines, so I made the code here do that too.

ahaldane commented:
I just pushed an extra commit with an alternate implementation of `void_repr`, which uses `formatter` rather than `repr` to show all the fields of the scalar. This will break a few more doctests, but I think I like it better. You can compare in the two last commits.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -859,3 +855,18 @@ def __call__(self, x):
+def _void_scalar_repr(x):
+    format_functions = []
+    for field_name in x.dtype.names:
+        format_function = _get_format_function(ravel(x[field_name]),
```
How about `format_function = repr`? I might be contradicting myself here, but I think that `repr(void) should not be reconfigurable - isn't that now the case for other scalar `repr`s?

eric-wieser replied:
Sorry, I missed your above comments - you already did this, and wanted me to compare.

eric-wieser commented:
The implementation of the second commit looks a little tidier, but I think we should be using `repr` inside `_void_repr` - for scalars, I think we should aim to keep the output constant no matter what the print setting sare.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -521,7 +515,9 @@ def array2string(a, max_line_width=None, precision=None,
 
-    if a.size == 0:
+    if style is not None and a.shape == ():
+        return style(a[()])
```
I think we can kill this `if` entirely now, and leave the `style` argument as obsolete?

ahaldane commented:
How do we want `x = np.array([(2., '2005-01-01', 3.)], dtype='f,M8[D],f')[0]` to be printed? If we use the formatters, we would get `(2., '2005-01-01', 3.)`, just like when printed as part of an ndarray. If we use the `repr`, we will get `(2.0, numpy.datetime64('2005-01-01'), 3.0)`. We will have a similar effect with unstructured void fields after #8981 because the void scalars will have a `void(` prefix.

I can see it both ways. 

The `formatter` version is more streamlined, and I like that the output of `print(arr)` and `print(arr[0])` will look the same, while there can be a big difference if we use `repr`. 

But I admit the `repr` version is more explicit since it shows the dtypes, which are lost in the `formatter` version. When we printed a full array using formatters, it was ok to skip the dtypes in the elements since the dtype is printed at the end of the array.

ahaldane commented:
Also, my original motivation for using `formatter` instead of `repr` was that void-scalars behave a lot more like arrays than all other scalars: They are writeable, sliceable, indexable, can be thought of as container-types, and are sequence-like. Therefore we might use `formatter` which is designed to print elements which are part of a sequence. Should `np.ones(2, dtype='f,f,f')` print so differently from `np.ones((2,3))`? Also, If the scalar contains subarrays, the output is already configurable by the print-options.

eric-wieser commented:
> Also, If the scalar contains subarrays, the output is already configurable by the print-options.

Is that true? I would have thought using `SubArrayFormat(repr)` (which results from my suggested change) would cause the print options to be ignored

eric-wieser commented:
> But I admit the repr version is more explicit 

Ultimately, I think this should be the goal of the scalar `repr`

> I like that the output of print(arr) and print(arr[0]) will look the same, while there can be a big difference if we use repr.

This is already untrue of floating point and datetime types, so I don't think it needs to be a goal.

> void-scalars behave a lot more like arrays than all other scalars

That's a reasonable argument, that I thought of too. But I think the important thing is that unlike arrays, `void.__repr__` does not show any dtype information, so needs to show that by repeatable (ie, not configurable) formatting

eric-wieser commented:
Either way, no point touching this till we pull the trigger on #9139

eric-wieser commented:
I've gone ahead and rebased this, and then added a single commit on the end with the `repr` behaviour - feel free to revert.

eric-wieser commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -634,6 +608,34 @@ static PyObject *
 static PyObject *
+voidtype_str(PyObject *self)
+{
+    if (PyDataType_HASFIELDS(((PyVoidScalarObject*)self)->descr)) {
```
Extra parens here

ahaldane replied:
No, I think we need them all... compiler errors for `if (PyDataType_HASFIELDS((PyVoidScalarObject*)self->descr))`

eric-wieser replied:
My bad, I was incorrectly reading `if ((PyDataType_HASFIELDS(...)))` here

eric-wieser commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -237,12 +237,27 @@ def test_formatter_reset(self):
+        if sys.version_info[0] >= 3:
+            assert_equal(str(np.array('café', np.unicode_)), 'café')
+            assert_equal(repr(np.array('café', np.unicode_)),
+                         "array('café',\n      dtype='<U4')")
```
This has an issue number, right?

ahaldane replied:
I don't think so... I think we came up with this one in #9143.

eric-wieser commented:
Actually, @mhvk [seems to think](https://github.com/numpy/numpy/pull/9143#issuecomment-302792838) that we don't want to change the void repr. Can you elaborate on that comment, @mhvk?

mhvk commented:
What we use in astropy quite a lot is structured arrays with multiple floats, and I liked very much that a scalar of such (say, 3 floats, for X, Y, Z) was typeset just like a one-dimensional array with length 3. (And, as noted in the comment you linked to, I already changed lots of tests in astropy to match the new behaviour, plus inserted code to ensure the same happened on numpy <1.12 -- https://github.com/astropy/astropy/blob/master/astropy/coordinates/representation.py#L42).

To be honest, though, I'm a bit lost as to what would change relative to 1.12 for scalar structured arrays.

eric-wieser commented:
> To be honest, though, I'm a bit lost as to what would change relative to 1.12 for scalar structured arrays.

```
>>> dt = np.dtype([('a', 'M8[D]'), ('b', float, (2,)), ('c', [('d', np.float16)])])
>>> x = np.array([('2017', [0.1, 1], (0.1,))], dt)
>>> x
....
>>> x[0]
...
```

gives on 1.12:
```
array([('2017-01-01', [ 0.1,  1. ], ( 0.09997559,))], 
      dtype=[('a', '<M8[D]'), ('b', '<f8', (2,)), ('c', [('d', '<f2')])])
('2017-01-01', [ 0.1,  1. ], ( 0.09997559,))
```

and in this patch on top of master

```
array([('2017-01-01', [0.1, 1. ], (0.09997559,))],
      dtype=[('a', '<M8[D]'), ('b', '<f8', (2,)), ('c', [('d', '<f2')])])
(numpy.datetime64('2017-01-01'), [0.10000000000000001, 1.0], (0.099976,))
```

ahaldane commented:
> > Also, If the scalar contains subarrays, the output is already configurable by the print-options.

> Is that true? I would have thought using `SubArrayFormat(repr)` (which results from my suggested change) would cause the print options to be ignored

OK, it does cause print options to be ignored. But I also think it's wrong to use the `repr` inside the subarray: Subarrays should look more or less like arrays. 

In other words, currently (using `repr`) we have this behavior:
```python
>>> np.zeros(1, dtype='f,(2,3)M8[D]')[0]
(0.0, [[numpy.datetime64('1970-01-01'), numpy.datetime64('1970-01-01'), numpy.datetime64('1970-01-01')], [numpy.datetime64('1970-01-01'), numpy.datetime64('1970-01-01'), numpy.datetime64('1970-01-01')]])
```
which doesn't seem right. 

If we want to be really consistent, perhaps we should print
```python
(0.0, np.array([['1970-01-01', '1970-01-01', '1970-01-01'], ['1970-01-01', '1970-01-01', '1970-01-01']], dtype=np.datetime64))
```

But I am not even sure I want that.. I might be most happy with something close to 1.13 notation.

ahaldane commented:
One more far out possibility: What if we modify the `void` scalar constructor to accept dtype, and then print the void scalar as ``np.void((0., 0.), dtype=[('f0', '<f4'), ('f1', '<f4')])``

mhvk commented:
I've been confusing myself no end for the last 30 minutes, so I ended up simply running the astropy tests. I get lots of doctest failures like the following:
```
040     >>> ICRS(ra=8.67*u.degree, dec=53.09*u.degree,
Differences (unified diff with -expected +actual):
    @@ -1,4 +1,4 @@
     <ICRS Coordinate: (ra, dec) in deg
    -    ( 8.67,  53.09)
    +    (8.6700000000000017, 53.090000000000011)
      (pm_ra_cosdec, pm_dec) in mas / yr
    -    ( 4.8, -15.16)>
    +    (4.7999999999999998, -15.16)>
```
Here, the spaces I'm not worried about, but the gratuitous final bit changes are, I think, quite annoying, even more so since one wouldn't have an easy way to control it.

Note that internally these use array scalars, so the discussion is slightly separate from the above. The main problem boils down to this:
```
In [17]: v = np.array((0.1, 2., 4.), dtype='f8,f8,f8')

In [18]: v
Out[18]: 
array((0.1, 2., 4.),
      dtype=[('f0', '<f8'), ('f1', '<f8'), ('f2', '<f8')])

In [19]: str(v)
Out[19]: '(0.10000000000000001, 2.0, 4.0)'
```
My own sense is that if `repr` is formatted for an array scalar, `str` should be too.

eric-wieser commented:
> My own sense is that if repr is formatted for an array scalar, str should be too.

I think the main problem here is that we need `str(str_array_scalar)` to give the raw string value, not a formatted value. Perhaps that should be special-cased at a higher level though.

mhvk commented:
I guess in part there is a different thing that bothers me:
```
In [48]: np.float64(0.1)
Out[48]: 0.10000000000000001

In [49]: np.float64(0.1).item()
Out[49]: 0.1
```
Why does `float64` typeset differently from a regular python `float`?

mhvk commented:
Bit more on track. Is it possible to agree that for array scalars, `repr` should stay as it is, since it is currently a logically consistent way to typeset all `ndarray`, independent of shape:
```
x = np.array([(2., '2005-01-01', 3.)], dtype='f,M8[D],f')
x
# array([(2., '2005-01-01', 3.)],
#       dtype=[('f0', '<f4'), ('f1', '<M8[D]'), ('f2', '<f4')])
x[0][...]
# array((2., '2005-01-01', 3.),
#       dtype=[('f0', '<f4'), ('f1', '<M8[D]'), ('f2', '<f4')])
```
If so, I think there is only one logical answer for what `str` should do: since for an array,
```
str(x)
# "[(2., '2005-01-01', 3.)]"
```
for an array scalar it should give the same except without the square brackets, i.e., just what `repr` uses after `array(`.

To me, for true scalars, I would prefer to not try to look like python items at all, i.e., just print `np.float64(1.)`, etc., but that ship probably has sailed a long time ago. But next best for `void` might be to just recursively `repr` for `repr` and recursively `str` for `str`.

mhvk commented:
@ahaldane - given what I just wrote, my preference definitely would be also to print void scalars as `void(..., dtype=...)`, i.e., that it would just look like the repr of an array scalar, but with `array` replaced by `void`. Unlike for `np.float64` and co, this might be OK, in not breaking expectations too much.

ahaldane commented:
I think we are in danger of getting carried away. Yes, probably it would be best in numpy 3.0 to prefix void structured scalars with `void(`. Probably, we would also want numpy scalar reprs to use the configurable `formatter` output, which would solve the different output precision of the scalars and 0d arrays.

One reason I like simply using `formatter` to print void-scalar elements, and not adding the `void(` prefix, (ie, like `(2., '2005-01-01', 3.)`) is its minimality: It is what 1.13 did (it used `formatter` to print the elements). I would rather leave it that way, than try to come up with some new behavior involving `repr` which I think is quite questionable.

Also, note that this whole issue of how to print void structured scalars can be split off from this PR, which was meant to fix 0d arrays, not void scalars. Why don't we leave it be with formatter here, and someone can open another PR if we want to change it.

ahaldane commented:
Just as a reminder, in numpy 1.13 structured-void-scalars were effectively printed by doing:
```python
arr = array(voidstructuredscalar)
format_functions = []
 for field_name in arr.dtype_.names:
        format_function = _get_format_function(data[field_name], **options)
        if dtype_[field_name].shape != ():
            format_function = SubArrayFormat(format_function)
        format_functions.append(format_function)
return StructureFormat(format_functions)(arr[0])
```

In this PR (without the `repr` commit at the end) we would be doing *exactly* the same, except without the conversion to 0d and back in the 1st and last line.

ahaldane commented on numpy/core/arrayprint.py:
```diff
@@ -869,7 +859,14 @@ def __call__(self, arr):
 
 class StructureFormat(object):
-    def __init__(self, format_functions):
+    def __init__(self, data, **options):
```
Is this too back-incompatible, or can we let the signature change slide? (Ie, is anyone besides us really calling StructureFormat?)

eric-wieser replied:
Why not just add `StructFormat.from_data`? I think the old signature was handy

ahaldane replied:
handy, yet never used, now.

In any case, good idea to avoid any compatibility issue, done.

mhvk commented:
@ahaldane - Definitely happy with this approach, and agreed that we can and should have further discussion, if any, on void scalars elsewhere.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -880,6 +884,13 @@ def __call__(self, x):
 
+# This function implements the repr for structured-void scalars. It is called
+# from the scalartypes.c.src code, and is placed here because it uses the
+# elementwise formatters above.
```
Any reason this isn't a docstring?

eric-wieser commented:
You're right, changing the scalar `repr` was never the point of this patch - I was getting distracted. I still think we want `str(0d)` to avoid `setprintfunction` entirely (perhaps handled in the C code), but this patch is still an improvement over what we had.

eric-wieser commented:
Can you add a test and "fixes" comment for #9345?

ahaldane commented:
Done (docstring) and Done (test/fixes).

eric-wieser commented on doc/release/1.14.0-notes.rst:
```diff
@@ -251,22 +251,40 @@ programs which open them with ``mmap``, especially on Linux where an
+scalars and higher dimension ndarrays.
+
+Now, the ``str`` of a 0d array prints like a numpy scalar using
+``str(a[()])`` and the ``repr`` prints like ndarrays using ``formatter(a[()])``,
```
`s/ndarrays/other arrays/` or `other dimensions of array

eric-wieser commented on doc/release/1.14.0-notes.rst:
```diff
@@ -251,22 +251,40 @@ programs which open them with ``mmap``, especially on Linux where an
+
+User-defined types now need to implement ``__str__`` and ``__repr__``
+---------------------------------------------------------------------
+Previously, user-defined types could fall back to a default implementation of
```
Should probably mention `np.generic` here

ahaldane replied:
what should it say? I'm unclear

eric-wieser reviewed (commented):
Few release note comments, but otherwise LGTM

eric-wieser commented:
> and that can wait for another PR.

Rethinking this. I'm happy with the fix here for recursion in `void.repr`, but I'm not too fond of reintroducing the `style` argument if the follow up PR is just going to deprecate it again.

I've made a competing fix to that problem at #9776. Thoughts?

eric-wieser commented:
I know I keep doing this, but I've split this PR in two - #9784 contains only the recursion fixing commit (the second one). Merging that should be enough to set the ball rolling on #8981.

ahaldane commented:
Rebased after #9784 so we can compare this and #9201.

eric-wieser commented:
I've just rebased #9201 too.

mhvk commented:
I must admit I'm not completely sure any more what the differences between the two PRs are, but I think both are a real improvement for `repr` and neither quite does what I would like for `str`, which is to follow `repr` and treat array scalars by default the same way as arrays. I think for structured arrays the case is [quite obvious](https://github.com/numpy/numpy/pull/9332#issuecomment-332275363) and then consistency suggests that the same should happen for other `dtype` (least surprise and all that). But perhaps I am missing some obvious downsides?

Given this, I think there are the following questions:
1. Should `str` by default use the formatters.
2. Should one keep `style` so that it can be used to do something else for array scalars (if not `None`; an argument against this is that almost certainly `style` was mostly used in the opposite sense before, to ensure scalars were typeset the same way... at least in astropy I know this is true).
3. Should we go even further and just pass all scalars through the same `array_print`? This seems reasonable at least for `void`.
4. Alternatively, should we always convert proper scalars to their python equivalent before converting so that we can get rid of really [silly formatting](https://github.com/numpy/numpy/pull/9332#issuecomment-332271948). (This would not work for `void`.)

I guess only (1) and (2) would have to be decided as part of the current pair of PRs.

eric-wieser commented:
Neither PR has any effect on `repr`

On Thu, Sep 28, 2017, 09:35 Marten van Kerkwijk <notifications@github.com>
wrote:

> I must admit I'm not completely sure any more what the differences between
> the two PRs are, but I think both are a real improvement for repr and
> neither quite does what I would like for str, which is to follow repr and
> treat array scalars by default the same way as arrays. I think for
> structured arrays the case is quite obvious
> <https://github.com/numpy/numpy/pull/9332#issuecomment-332275363> and
> then consistency suggests that the same should happen for other dtype
> (least surprise and all that). But perhaps I am missing some obvious
> downsides?
>
> Given this, I think there are the following questions:
>
>    1. Should str by default use the formatters.
>    2. Should one keep style so that it can be used to do something else
>    for array scalars (if not None; an argument against this is that
>    almost certainly style was mostly used in the opposite sense before,
>    to ensure scalars were typeset the same way... at least in astropy I know
>    this is true).
>    3. Should we go even further and just pass all scalars through the
>    same array_print? This seems reasonable at least for void.
>    4. Alternatively, should we always convert proper scalars to their
>    python equivalent before converting so that we can get rid of really silly
>    formatting
>    <https://github.com/numpy/numpy/pull/9332#issuecomment-332271948>.
>    (This would not work for void.)
>
> I guess only (1) and (2) would have to be decided as part of the current
> pair of PRs.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/9332#issuecomment-332892819>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ9LNQoyfVoHkSHWTVMyjEQzFcZqQWhks5sm8rEgaJpZM4OJwm0>
> .
>

eric-wieser commented:
>  I think both are a real improvement for repr

Neither PR has any effect on `repr`

>  I would like for str, which is to follow repr and treat array scalars by default the same way as arrays. 

I believe this is the current state of master

> I must admit I'm not completely sure any more what the differences between the two PRs are,

This one allows the result of `str(array('string'))` to be changed via `np.set_string_function`, the other one does not. This one also keeps the `style` argument from 1.13, the other one leaves it out like master.

mhvk commented:
Thanks for the explanation; I clearly had gotten lost!

Since I prefer not changing array scalars to normal scalars, my tendency would be to keep things as they are in current master and merge neither... (but if backward compatibility is of some concern, I would also be happy to with a `style` that acts on the array scalar directly, as it does in 1.13, but which is turned off by default).

ahaldane commented:
We should choose one of them because current master behavior fails the test `assert_equal(str(np.array('café', np.unicode_)), 'café')`, which I think we found that some people depend on.

ahaldane commented:
I don't see a huge difference between the two PRs, they are both acceptable. It's hard to decide which one is a smidgeon better. The question involves details in how 0d arrays work in unusual user-customization cases which aren't super important.

The question comes down to: Should we make `str(0d)` customizable by `np.set_string_function`. If yes, then we need to keep the style argument and use this PR. If no, we want #9784.

In the long-term I think that 0d arrays *should* be customable by `set_string_function`, since maybe we want ultimately remove scalars and replace them by 0d arrays. Then all numpy output would be customizable.

I think I favor this PR because 1. it is most back-compatible with 1.13 (keeps the style argument), and 2. I slightly prefer if 0d array str is customizable.

ahaldane commented:
Eric raised a question about unicode in #9201. I didn't exaclty understand the point, but I do notice that both of our PRs fail on `str(np.array(u'café', np.unicode_))`, while 1.13 did not.

mhvk commented:
Ah, yes, on current master (for both
```
str(np.array('café'))
# "'café'"
```
The problem is the extra quotes, which are logical for a multi-element array, but less so for a singleton. An alternative solution might be to treat this as a special case in the formatter (just as we do for `True` in bool).

ahaldane commented:
@mhvk the difficulty is we want the quotes in the repr, but not the str:
```python
>>> repr(np.array('café'))
"array('café',\n      dtype='<U4')"
>>> str(np.array('café'))
'café'
```
so I don't think something in the formatter will help, since we don't know at that point whether we are in the str or repr case.

Also my last comment was a red-herring - that example never worked in python2.

mhvk commented:
Hmm, yes, I see the problem now. Possibly the solution is then to special-case string/unicode array scalars only?

Somewhat separately, for this PR, if we keep `style`, I don't see a reason to feed it the scalar rather than the array scalar (might break code for no good reason, given that scalars do not have, e.g., a functional `.view`).

ahaldane commented:
> Somewhat separately, for this PR, if we keep style, I don't see a reason to feed it the scalar rather than the array scalar

Isn't that recursive? We can't implement `str(zdarr)` by returning `str(zdarr)` ! That's why we return `str(zdarr[()])` instead.

And as for special-casing only unicode types, possible, though a bit hacky. Also note other types are changed too. Eg, `str(array(1.0))` gives `1.` with master, but gives `1.0` with this PR and with 1.13.

ahaldane committed:
MAINT: create a "legacy" print mode to account for whitespace and 0d changes

Fixes #9804

ahaldane commented:
I added a commit here following discussion in #9804.

This adds a new "legacy" option to array2string (which replaces the old `sign='legacy'` option, which is now removed). Now when this legacy mode is enabled, the old whitespace behavior, and the old 0d behavior, are used.

ahaldane commented:
I put the legacy option in the **kwargs so that it is easier to remove later down the line.

ahaldane commented:
While we let this sit to consider the options, I'm going to focus my current efforts, given limited time, on getting the `LongFloatFormatter` changes done.

eric-wieser commented:
@ahaldane: Can I suggest you do the long float work on `master` to avoid it becoming tangled in this issue?

ahaldane commented:
Good idea, will try.

mhvk reviewed (approved):
This is quite a bit nicer, not mixing the meaning of `sign` and legacy behaviour. Total nitpicks:
- I think normally one uses `**kwargs` rather than `**kwarg`.
- Should the docstrings state that this is a keyword-only argument? (since it is at the end and the signature already indicates it, perhaps not doing so is fine).

mhvk commented:
Looking at this again: I'm happy with this PR as is (modulo nitpicks), and think we should move the discussion of the `str(array-scalar)` to a separate issue (since nothing here really addresses it, I think, except that the text of the changelog is adapted to reflect the current reality; or am I wrong??).

I do think a separate PR would be useful, so that `str(array-scalar)` case can produce the same as `repr()` has inside the `array()` except for string/unicode (where I agree the quotes need to be gone).

eric-wieser commented:
@mhvk: pretty sure this patch handles `str(arr0d)`, unless ahaldane removed it.

I'm still not a fan of adding back the ugly `style` argument

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -503,18 +518,25 @@ def array2string(a, max_line_width=None, precision=None,
+        return style(a.item())
+
+    if style is not None and a.shape == ():
+        return style(a[()])
```
I don't think this is recursion-safe for `class subclass(np.ndarray): pass; a = subclass()` - `[()]` doesn't return a scalar for subclasses.

ahaldane replied:
it seems to return a scalar here...
```python
[1]: class sub(np.ndarray):
...:     pass
...: 
[2]: a = np.array(1.0).view(sub)
[3]: a[()]
1.0
[4]: type(a[()])
numpy.float64
```

mhvk replied:
Indeed, that is the default. I'm not sure how this would work with class which like `Quantity` explicitly circumvent this by overriding `__getitem__` (since it obviously would be horrible if just looking at an individual item would lose the unit... I think I may have mentioned I disliked the scalar classes? ;-).  

Now, since `Quantity` overrides `__str__` too, this is not an issue for it, but it adds to @eric-wieser's point that perhaps one should not count on `[()]` returning a different class. That said, in this situation the user is providing their own `style`, so at some level it is up to them to ensure it is a function that doesn't just call back.

ahaldane replied:
Well if we merge #9201 this problem disappears!

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -503,18 +518,25 @@ def array2string(a, max_line_width=None, precision=None,
+    if options['legacy'] and a.shape == () and not a.dtype.names:
+        if style is None:
+            style = repr
+        return style(a.item())
```
I think I'd prefer that the `style` argument remain deprecated _unless_ `legacy=True`.

mhvk replied:
That seems like a good idea.

ahaldane replied:
If you write that up in #9201 I will consider it, though I still slightly disagree since it seems like an almost useless back-compatibility break.

Also, at first glance it's unclear how to do that, since to get rid of `style` you need to hardcode the 0d `str` so it is unaffected by `np.set_printoptions`... but then `array2string` would never get called, nor its legacy branch. But probably there is a way.

eric-wieser reviewed (dismissed):
Drawing attention to [this comment](https://github.com/numpy/numpy/pull/9332#discussion_r142721150)

ahaldane commented:
After looking over this and #9201 again, how about this:

I think the hardcoding of `str(0darr)` in #9201 might be OK, because it might be what we want long-term once scalars become 0d arrays. If not, it is easy to change and probably doesn't affect almost anyone.

So I think we could merge #9201, and then I would still like to rebase this on top of that because I like moving the `legacy` option to its own parameter.

mhvk commented:
I'd still prefer to move the other way, i.e., ensure array scalars look as much as possible as regular arrays, including being affected by `set_print_options`, and without any translations to scalars. (Just noticed that #9883 already does that for `__format__`, though since in that case an explicit format is given, it doesn't matter as much).

mhvk commented:
As it has been a while, and I've been too busy with other things, it would be useful for me at least to be reminded what the main rationale for converting to scalar actually is. What I do recall is the reasonable case that one wants `str(array('string')) == 'string'` to evaluate as `True`. Can this be as simple as special-casing just the string dtypes? I do feel that numerical and structured array scalars should just set consistently with other numerical arrays (i.e., obey `set_print_options`).

So, my suggestion would be to translate to scalar only in special cases where there is a very specific reason.

mhvk commented:
p.s.  As for `style` - I'd go with @eric-wieser's suggestion of only using it for legacy printing.

ahaldane committed:
MAINT: Move 0d special casing to array_str and array_repr

ahaldane commented:
I'm slow in responding because I myself have trouble remembering all the devilish details involved here. Let me try...

A key part of the puzzle is that in `array2string` we normally have no way of knowing whether we are printing a `repr` or a `str`, and in almost all cases it doesn't matter. Normally `array_str` just returns `array2string(arr)` and `array_repr` returns `"array(" + array2string(arr) + ")"` and everything was mostly OK. 

The exception is 0d arrays. In this case users expect `array_str` to return `str(arr[()])` and `array_repr` to return `"array(" + repr(arr[()]) + ")"`, in other words we can't drop in a single `array2string` return value like before, it needs to be different for the two cases.

One way or another, we will need to special-case 0d arrays in order to maintain this distinction between `str` and `repr`. There are a few ways we are considering doing this:

  1. The solution in previous numpy versions, and in this PR, has been to have an extra `style` argument to `np.array2string` just for 0d arrays, which `array_str` fills in with `str`, and `array_repr` with `repr`. In other words, the special case was in `array2string`, which is user-configurable.
  2. In #9201, the special case has been moved deep into the C `array_str` code so it is not user-configurable. Now `str(0d)` always exactly returns `str(arr[()])`. We would deprecate the `style` argument of `np.array2string`. But as Eric suggested, in this case we could still leave `style` active in legacy mode only.

ahaldane commented:
(Note that it does not make sense to only use `style` in legacy mode in this PR - this PR makes full use of the style argument even in non-legacy mode)

(In current master, I mistakenly got rid of the style argument without realizing I would break user's expectations. So this PR is just adding back in the style argument that was there before).

(To add some confusion, this PR also fixes some other problems with the same code at the same time, eg it makes `array_repr` return `formatter(arr[()])` to be like other ndarrays, but this change is uncontroversial and tangent to the problem we are discussing)

ahaldane commented:
To further bend our minds, I've added a commit with yet another option. (and rebased on master) 

Now, this PR *allows* the result of `str(array('string'))` to be changed via `np.set_string_function` (unlike #9201) but still deprecates/ignores the `style` argument from `array2string` (like #9201), but still allows `style` to be functional in `legacy` mode.

This is achieved by moving the special-casing of 0d arrays into `array_str` and `array_repr`.

mhvk commented:
@ahaldane - Thanks for the nice summary. I like the last commit in that it seems to put the choices where they belong. This makes them much easier to discuss, and also makes `array2string` cleaner, so I'm very much in favour!

It also means my comment can be much more focused: when should an array scalar be converted to a normal scalar. We all seem happy to never do that for `repr`. For `str`, I think we also all agree that we should generally do it for dtype `S` and `U` (or at least that we don't want quotes around a string array scalar). I would argue that
1. We should definitely *not* do it for structured `dtype` (i.e., dtype `f8,f8,f8` should be set just like a 3-element array);
2. I see no reason to do it for any numerical type, it seems reasonable that `set_print_options` acts on array scalars as well;
3. I basically see no point for `datetime` either.

So, I guess my suggestion would be to special-case only `dtype.kind in 'SU'` for non-legacy mode...

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -541,23 +554,31 @@ def array2string(a, max_line_width=None, precision=None,
+    elif not (style is np._NoValue):
+        warnings.warn("'style' argument is deprecated and no longer functional"
+                      " except in 'legacy' mode",
+                      DeprecationWarning, stacklevel=3)
```
This logic isn't right, it will warn for `array2str([1], style=str, legacy=True)` - you need an outer `if` switching on `legacy`.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -1105,7 +1127,9 @@ def array_repr(arr, max_line_width=None, precision=None, suppress_small=None):
 
-    if arr.size > 0 or arr.shape == (0,):
+    if _format_options['legacy'] and arr.shape == () and not arr.dtype.names:
+        lst = repr(arr[()])
```
Wouldn't `arr.item()` match the legacy behaviour?

eric-wieser commented:
I like the look of this revision too. Seem to be a couple bugs.

> I would argue that ...

Note that this would be confusing when used with #7978:

```
>>> x = np.array(0)
>>> "value is {}".format(x)
"value is 0"
>>> "value is %s" % x
"value is array(0)"
```

My argument for the flattening in `str` is that it means a user doesn't need to care about 0d arrays vs scalars when trying to `format` / `%` a string.

mhvk commented:
@eric-wieser - just to be sure: my suggestion would also not print out `array(0)`. Rather, it would do:
```
"value is {!s}".format(np.array(1.0123456789))
# 'value is 1.01234568'
np.set_printoptions(precision=5)
"value is {!s}".format(np.array(1.0123456789))
#  'value is 1.01235'
```
But you're right that given #9883, the precision would be ignored without the `!s`:
```
"value is {}".format(np.array(1.0123456789))
# 'value is 1.0123456789'
```
As you might have guessed, I'd ideally solve this! I note that structured dtypes do work right (I think because `void` are passed through `array2string`):
```
x = np.array((1.123456789,1.2,1.3), dtype='f8,f8,f8')
"value is {}".format(x)
# 'value is (1.12346, 1.2, 1.3)'
```

It would seem there two possibilities:
1. Do the conversion to scalar only if format is not an empty string (i.e. make #9883 more specific to the case it was aiming to solve).
2. Make numerical scalars behave more like arrays, passing them through `array2string` (as happens already with `void`)

mhvk commented:
p.s. I should add that for *this* PR, as far as I'm concerned, my only suggestion is to special-case only what we really want to special-case, which is string scalars. I made a PR to this one that just removes quotes on each side for those.

mhvk commented:
See https://github.com/ahaldane/numpy/pull/1

ahaldane commented:
I actually somewhat dislike truncating floats if they appear on their own. Inside a `"array(" + x + ")"`, or inside `[]` its fine, but on its own it could be confusing because it looks like a scalar but isn't. Scalars don't truncate, so with your suggestion with a `str(0d)` you might think you got a unique representation, but didn't.

So I still favor that `str(0d)` is printed exactly like scalars. Scalar floats print in unambiguous full precision. In general scalars and 0ds are printed on their own so we don't have limited space like when printing an array.

mhvk commented:
OK, that is a good argument as well. And since with your current implementation I can override this if I wish by using `np.set_string_function`, I'm happy to just leave it as is (modulo the comments @eric-wieser made). It is a great improvement as is!

ahaldane committed:
DOC: Update release notes about legacy mode and 0d arrays

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -1164,6 +1184,11 @@ def array_str(a, max_line_width=None, precision=None, suppress_small=None):
+    if a.shape == () and not a.dtype.names:
+        if _format_options['legacy']:
+            return str(a.item())
+        return str(a[()])
```
What effect is guarding this under `a.dtype.names` in the non-legacy case having?

eric-wieser replied:
I'd perhaps be inclined to write this as the slightly longer

```
if _format_options['legacy'] and a.shape == () and not a.dtype.names:
    return str(a.item())

# actually justify this behavior in a comment here, where its a feature, not deprecated behaviour
if a.shape == () and not a.dtype.names:
    return str(a[()])
```
just to separate the legacy conditions from the rest

ahaldane replied:
Yeah good idea for clarity.

I had the same thought about `a.dtype.names` when writing this. I did confirm for myself that in 1.13 the `a.dtype.names` case was special-cased to use the `formatter` to represent the value, i.e. it should go through the `array2string` code, which uses `formatter`.

ahaldane replied:
Hmm but actualy, here is the implication of excluding `a.dtype.names` in the non-legacy case.

If we do `if a.shape == () and not a.dtype.names:`, then the eventual output is the output of `StructureFormat`, but with the `precision` specified as argument to `array_str`, but all other settings from `__format_options`. I am not sure the `precision` argument is ever used.

If we only do `if a.shape == ()`, then the eventual output is still the output of `StructureFormat`, but with  all settings, including `precision`, from `__format_options`. We go through a different path through `void_scalar_repr` to get there but the result is almost identical.

I'm going to remove the `a.dtype.names` test in the non-legacy case, just to make the code simpler. I don't think it has any real effect. I'll leave it in legacy mode since  it makes a difference there,

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -541,23 +553,30 @@ def array2string(a, max_line_width=None, precision=None,
+    if options['legacy']:
+        if a.shape == () and not a.dtype.names:
+            return style(a.item())
+    elif not (style is np._NoValue):
```
nit: `style is not np._NoValue`

eric-wieser commented on doc/release/1.14.0-notes.rst:
```diff
@@ -256,6 +269,16 @@ Chebyshev points of the first kind. A new ``Chebyshev.interpolate`` class
+character for positive values. The new default is '-'.
+
+This new default changes the float output relative to numpy 1.13. The old
+behavior can be obtained in "legacy" printing mode, see compatibility notes.
```
nit: add the word "above" here, and in the other place you refer to it.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -541,23 +553,30 @@ def array2string(a, max_line_width=None, precision=None,
+        if a.shape == () and not a.dtype.names:
+            return style(a.item())
+    elif style is not np._NoValue:
+        warnings.warn("'style' argument is deprecated and no longer functional"
```
You lost the deprecation date comment here

eric-wieser commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -302,11 +317,13 @@ def test_sign_spacing(self):
         assert_equal(repr(b), 'array([+1.234e+09])')
 
-        np.set_printoptions(sign='legacy')
+        np.set_printoptions(legacy=True)
```
Would be nice to have a test that `style` is not deprecated in legacy mode.

charris commented:
Marking this for 1.14 release. If there are more unmerged PRs involving printing/formatting, it would be helpful to also set their milestones.

ahaldane commented:
@charris, I just went through and added the last 2 to 1.14

charris commented:
Great, thanks Allan. I'm going to start pushing off most of the other PRs to 1.15, but I want to make sure all the printing/formatting changes get in.

ahaldane commented:
I think everything is done.

I'd like to merge this one before rebasing #8981 because of the dependencies.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -541,23 +553,31 @@ def array2string(a, max_line_width=None, precision=None,
+        if a.shape == () and not a.dtype.names:
+            return style(a.item())
+    elif style is not np._NoValue:
+        # Deprecation 11-9-2017  v1.14
```
Can we make this `2017-11-09` please (like we do elsewhere)? As a European, this format is silly ;)

(yes, I realized that I missed it the first round)

eric-wieser reviewed (approved):
All looks good to me! I've rebased #9201 with just the unicode changes.

One tiny nit about date formats in deprecation warning.

charris commented:
Sounds like folks are happy with this, so in it goes. Thanks Allan, and thanks to the rest of the print formatting crew.
